[Security Solution] JSON diffs test plan (#175958)
elastic/kibana: x-pack/plugins/security_solution/docs/testing/test_plans/detection_response/prebuilt_rules/installation_and_upgrade.md

Then all X rules available for upgrade should be displayed in the table
When user opens the rule preview for the 1st rule
Then the preview should open

> And the "Update" tab should be active

Reviewer: nit: "Updates"

TODO: add scenarios https://github.com/elastic/kibana/issues/166215

### Rule upgrade workflow: viewing rule changes before upgrading

> #### **Scenario: User can see changes in a side-by-side view**

Reviewer 2: Nit:

```suggestion
#### **Scenario: User can see changes in a side-by-side JSON diff view**
```

**Automation**: 1 Cypress e2e test

```Gherkin
Given X prebuilt rules are installed in Kibana
And for Y of these rules new versions are available
When user opens the Rule Updates table and selects a rule
Then the upgrade preview should open

> And rule changes should be displayed in a two-column view

Reviewer 2: ```suggestion
And rule changes should be displayed in a two-column JSON diff view
```

And correct rule version numbers should be displayed in their respective columns
When the user selects another rule without closing the preview
Then the preview should display the changes for the newly selected rule
```

#### **Scenario: User can see precisely how property values would change after upgrade**

**Automation**: 1 UI component-level test

```Gherkin
Given a rule preview with rule changes is open
Then each line of <column> that was <change_type> should have <bg_color> background
And marked with <line_badge> badge

> And each changed word in <column> should be highlighted with <accent_color>

Reviewer: I cannot recall now, but we are only using only WORD diffing for the whole JSON, right? No need to test other uses cases, like character diffing.

Author: No, at this stage it's by word only.

Examples:
| change_type | column       | bg_color         | accent_color         | line_badge |

> | updated     | Base version | removed_bg_color | removed_accent_color | -          |

Reviewer: `Base Version`

According to the discussion we had yesterday, I think it's pretty safe to assume that this will need to be changes to `Current Version` or `Current Rule`.

If we are bringing that uo in today's Product Meeting, we can ask, and then change the label as part of this PR and update the wording in the test plan.

| updated     | Update       | added_bg_color   | added_accent_color   | +          |
| removed     | Base version | removed_bg_color | none                 | -          |
| removed     | Update       | none             | none                 | none       |
| added       | Base version | none             | none                 | none       |
| added       | Update       | added_bg_color   | none                 | +          |
```

#### **Scenario: Rule actions and exception lists should not be shown as modified**

**Automation**: 1 UI component-level test

```Gherkin
Given a prebuilt rule is installed in Kibana

> And this rule doesn't have any actions or an exception list

Reviewer: nit: `And this rule doesn't have any actions or an exception list`

I think this might be a little bit confusing with the line that comes next to it. Maybe reformulate it to `And the base version of this rule doesn't have any actions or an exception lists`

And a user has set up actions and an exception list for this rule
And this rule has an update available
And the update doesn't define any actions or an exception list
When a user opens the upgrade preview for this rule
Then the preview should open

> And there should not be any modifications to rule's actions or exception list

Reviewer 2: Nit:
```suggestion
And the JSON diff shouldn't show any modifications to rule's actions or exception list
```

```

> #### **Scenario: Stateful properties should not be included in preview**

Author: My plan was to keep technical and dynamic properties separate, because I thought it would make sense to test dynamic properties using some kind of E2E test - install a rule, execute it once, then see if update lists only the properties we expect. Keep it more real-life since it relates to runtime changes.

And as for technical properties, we could just test them with something like react-testing-library and two harcoded rule objects. What do you think?

> #### **Scenario: Stateful properties should not be included in preview**

Reviewer 2: @nikitaindik sounds good to me 👍

**Automation**: 1 Cypress E2E test

```Gherkin
Given a prebuilt rule is installed in Kibana
And this rule is disabled by default
And a user has enabled this rule
And this rule executed at least once
And this rule has an update available
When user opens the upgrade preview
Then the preview should open

> And there should not be any changes related to "execution_summary" and "enabled" properties

Reviewer 2: Nit:
```suggestion
And the JSON diff shouldn't show any <property> properties on both sides

Examples:
| property          |
| execution_summary |
| enabled           |
```
